clarity/src/vm/functions/database.rs

```diff
@@ -924,3 +947,188 @@ pub fn special_get_burn_block_info(
+
+    Value::some(result)
+}
+
+/// Handles the function `get-tenure-info?` special function.
+/// Interprets `args` as variables `[property-name, block-height]`, and returns
+/// a property value determined by `property-name`:
+/// - `time` returns the burn block time for the tenure of which `block-height` is a part
+/// - `vrf-seed` returns the VRF seed for the tenure of which `block-height` is a part
+/// - `burnchain-header-hash` returns header hash of the burnchain block corresponding to the tenure of which `block-height` is a part
+/// - `miner-address` returns the address of the principal that mined the tenure of which `block-height` is a part
+/// - `miner-spend-winner` returns the number of satoshis spent by the winning miner for the tenure of which `block-height` is a part
+/// - `miner-spend-total` returns the total number of satoshis spent by all miners for the tenure of which `block-height` is a part
+/// - `block-reward` returns the block reward for the tenure of which `block-height` is a part
+///
+/// # Errors:
+/// - CheckErrors::IncorrectArgumentCount if there aren't 2 arguments.
+/// - CheckErrors::GetTenureInfoExpectPropertyName if `args[0]` isn't a ClarityName.
+/// - CheckErrors::NoSuchTenureInfoProperty if `args[0]` isn't a TenureInfoProperty.
+/// - CheckErrors::TypeValueError if `args[1]` isn't a `uint`.
+pub fn special_get_tenure_info(
+    args: &[SymbolicExpression],
+    env: &mut Environment,
+    context: &LocalContext,
+) -> Result<Value> {
+    // (get-tenure-info? property-name block-height-uint)
+    runtime_cost(ClarityCostFunction::BlockInfo, env, 0)?;
+
+    check_argument_count(2, args)?;
+
+    // Handle the block property name input arg.
+    let property_name = args[0]
+        .match_atom()
+        .ok_or(CheckErrors::GetTenureInfoExpectPropertyName)?;
+
+    let block_info_prop = TenureInfoProperty::lookup_by_name(property_name)
+        .ok_or(CheckErrors::GetTenureInfoExpectPropertyName)?;
+
+    // Handle the block-height input arg.
+    let height_eval = eval(&args[1], env, context)?;
+    let height_value = match height_eval {
+        Value::UInt(result) => Ok(result),
+        x => Err(CheckErrors::TypeValueError(TypeSignature::UIntType, x)),
+    }?;
+
+    let Ok(height_value) = u32::try_from(height_value) else {
+        return Ok(Value::none());
+    };
+
+    let current_height = env.global_context.database.get_current_block_height();
+    if height_value >= current_height {
+        return Ok(Value::none());
+    }
+
+    let result = match block_info_prop {
+        TenureInfoProperty::Time => {
+            let block_time = env
+                .global_context
+                .database
+                .get_burn_block_time(height_value)?;
+            Value::UInt(u128::from(block_time))
+        }
+        TenureInfoProperty::VrfSeed => {
+            let vrf_seed = env
+                .global_context
+                .database
+                .get_block_vrf_seed(height_value)?;
```

Comment:
This seems broken. Only tenure-change blocks have VRF proofs in the `nakamoto_block_headers` table.  Furthermore, the `get_vrf_seed_for_block()` function in `HeadersDB` appears broken for Nakamoto blocks, because the column isn't called `"proof"` in `nakamoto_block_headers`; it's now called `"vrf_proof"`.

Reply:
Okay, I see. This means that it was already broken prior to this change in a call to `get-block-info?`. I will fix it and check the other properties for similar issues.

Reply:
Right.  Turns out, most (all?) of these fields cannot be accessed from anything other than the tenure-start block.  Specifically:

* The `HeadersDB` implementations for Stacks chain state, when handling queries for `miner-address`, `miner-spend-total`, and `miner-spend-winner`, will all look in the `payments` table.  But, this table only contains rows pertaining to `BlockFound` tenure-change blocks in Nakamoto.  If you passed in an index block hash for a non-tenure-change Nakamoto block, your query would return no rows.

* `vrf-seed` is broken in two ways, per the above -- (1) the column name is different in Nakamoto (`vrf_proof`, not `proof`), which precludes using the `get_stacks_header_column()` function in `src/clarity_vm/mod.rs`, and (2) the VRF proof (and thus the seed, which is derived from it) is only defined for Nakamoto blocks which have a `Coinbase` transaction.  If you passed an index block hash for a non-coinbase Nakamoto block, you'd get no rows.

* `block-reward` is only defined for Nakamoto blocks with a coinbase, so it's broken for the same reason the `vrf-seed` is broken.

If `get-tenure-info` is going to take a Stacks block height, then it's going to need to resolve the identified Stacks block to its tenure-start block before issuing the above queries.

Reply:
Is there a clean way to get the epoch here? The simplest option is to check if we got `None` and then try again with `vrf_proof` for the column name, but if there is a better way to cheaply find our epoch or burn block height here, that seems better.

Reply:
> Turns out, most (all?) of these fields cannot be accessed from anything other than the tenure-start block

Okay, I didn't realize that. I'll figure out a way to get the tenure start block based on the given stacks height.

Reply:
Having read through this PR, there's basically zero integration test coverage of `get-stacks-block-info?` and `get-tenure-info?`.  These bugs could have been caught if you did something like this:

1. Construct a Nakamoto chain with e.g. 10 blocks per tenure
2. In each Nakamoto block, deploy a smart contract with this code body:

```clarity
(print {
   "stacks-block-height": (- stacks-block-height u1),
   "miner-address": (get-tenure-info? (- stacks-block-height u1) miner-address),
   "miner-spend-total": (get-tenure-info? (- stacks-block-height u1) miner-spend-total),
   "miner-spend-winner": (get-tenure-info? (- stacks-block-height u1) miner-spend-winner),
   "vrf-seed": (get-tenure-info? (- stacks-block-height u1) vrf-seed),
   "block-reward": (get-tenure-info? (- stacks-block-height u1) block-reward),
   "tenure-time": (get-tenure-info? (- stacks-block-height u1) time),
   "burnchain-header-hash": (get-tenure-info? (- stacks-block-height u1) burnchain-header-hash)
   "block-time": (get-stacks-block-info? (- stacks-block-height u1) time),
   "header-hash": (get-stacks-block-info? (- stacks-block-height u1) header-hash),
   "id-header-hash": (get-stacks-block-info? (- stacks-block-height u1) id-header-hash)
})
```
3. Capture each `print` event from each Nakamoto block
4. Compare each `print` event's tuple to what you'd expect to see in that Nakamoto block.

You can do something similar for the Stacks 2.x blocks leading up to Nakamoto.

Reply:
Agreed. I am already in the process of writing this test.